MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups

```
java.lang.AssertionError: expected:<10000> but was:<9840>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:834)
    at org.junit.Assert.assertEquals(Assert.java:645)
    at org.junit.Assert.assertEquals(Assert.java:631)
    at com.hazelcast.map.impl.mapstore.MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups(MapLoaderFailoverTest.java:164)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-OracleJDK1.6/com.hazelcast$hazelcast/881/testReport/junit/com.hazelcast.map.impl.mapstore/MapLoaderFailoverTest/testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups/

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-OracleJDK1.7/com.hazelcast$hazelcast/454/testReport/junit/com.hazelcast.map.impl.mapstore/MapLoaderFailoverTest/testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups/

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/MapKeyLoader.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/MapKeyLoader.java
@@ -21,6 +21,7 @@ import com.hazelcast.core.IFunction;
 import com.hazelcast.core.MapLoader;
 import com.hazelcast.core.Member;
 import com.hazelcast.internal.cluster.ClusterService;
+import com.hazelcast.logging.ILogger;
 import com.hazelcast.map.impl.mapstore.MapStoreContext;
 import com.hazelcast.map.impl.operation.LoadStatusOperation;
 import com.hazelcast.map.impl.operation.LoadStatusOperationFactory;
@@ -72,6 +73,8 @@ public class MapKeyLoader {
 
     private static final long LOADING_TRIGGER_DELAY = SECONDS.toMillis(5);
     private static final long KEY_DISTRIBUTION_TIMEOUT_MINUTES = 15;
+
+    private ILogger logger;
 
     private String mapName;
     private OperationService opService;
@@ -132,6 +135,7 @@ public class MapKeyLoader {
         this.clusterService = clusterService;
         this.toData = serialize;
         this.execService = execService;
+        this.logger = getLogger(MapKeyLoader.class);
     }
 
     public Future startInitialLoad(MapStoreContext mapStoreContext, int partitionId) {
@@ -142,6 +146,10 @@ public class MapKeyLoader {
 
         role.nextOrStay(newRole);
         state.next(State.LOADING);
+
+        if (logger.isFinestEnabled()) {
+            logger.finest("startInitialLoad invoked " + getStateMessage());
+        }
 
         switch (newRole) {
             case SENDER:
@@ -274,6 +282,10 @@ public class MapKeyLoader {
     }
 
     private void sendKeysInBatches(MapStoreContext mapStoreContext, boolean replaceExistingValues) throws Exception {
+
+        if (logger.isFinestEnabled()) {
+            logger.finest("sendKeysInBatches invoked " + getStateMessage());
+        }
 
         int clusterSize = partitionService.getMemberPartitionsMap().size();
         Iterator<Object> keys = null;
@@ -438,5 +450,16 @@ public class MapKeyLoader {
 
     public void onKeyLoad(ExecutionCallback<Boolean> callback) {
         loadFinished.andThen(callback, execService.getExecutor(MAP_LOAD_ALL_KEYS_EXECUTOR));
+    }
+
+    public void promoteToLoadedOnMigration() {
+        // The state machine cannot skip states so we need to promote to loaded step by step
+        state.next(State.LOADING);
+        state.next(State.LOADED);
+    }
+
+    private String getStateMessage() {
+        return "on partitionId=" + partitionId + " on " + clusterService.getThisAddress() + " role=" + role
+                + " state=" + state;
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/MapMigrationAwareService.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/MapMigrationAwareService.java
@@ -21,13 +21,13 @@ import com.hazelcast.map.impl.record.Record;
 import com.hazelcast.map.impl.record.Records;
 import com.hazelcast.map.impl.recordstore.RecordStore;
 import com.hazelcast.nio.serialization.Data;
-import com.hazelcast.spi.partition.MigrationEndpoint;
 import com.hazelcast.query.impl.Indexes;
 import com.hazelcast.query.impl.QueryableEntry;
 import com.hazelcast.spi.MigrationAwareService;
 import com.hazelcast.spi.Operation;
 import com.hazelcast.spi.PartitionMigrationEvent;
 import com.hazelcast.spi.PartitionReplicationEvent;
+import com.hazelcast.spi.partition.MigrationEndpoint;
 import com.hazelcast.spi.serialization.SerializationService;
 import com.hazelcast.util.Clock;
 
@@ -68,6 +68,12 @@ class MapMigrationAwareService implements MigrationAwareService {
         migrateIndex(event);
         if (event.getMigrationEndpoint() == MigrationEndpoint.SOURCE) {
             clearMapsHavingLesserBackupCountThan(event.getPartitionId(), event.getNewReplicaIndex());
+        }
+        PartitionContainer partitionContainer = mapServiceContext.getPartitionContainer(event.getPartitionId());
+        for (RecordStore recordStore : partitionContainer.getAllRecordStores()) {
+            // in case the record store has been created without loading during migration trigger again
+            // if loading has been already started this call will do nothing
+            recordStore.startLoading();
         }
         mapServiceContext.reloadOwnedPartitions();
     }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContext.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContext.java
@@ -106,6 +106,8 @@ public interface MapServiceContext extends MapServiceContextInterceptorSupport, 
 
     RecordStore getRecordStore(int partitionId, String mapName);
 
+    RecordStore getRecordStore(int partitionId, String mapName, boolean skipLoadingOnCreate);
+
     RecordStore getExistingRecordStore(int partitionId, String mapName);
 
     Collection<Integer> getOwnedPartitions();

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContextImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContextImpl.java
@@ -300,6 +300,11 @@ class MapServiceContextImpl implements MapServiceContext {
     }
 
     @Override
+    public RecordStore getRecordStore(int partitionId, String mapName, boolean skipLoadingOnCreate) {
+        return getPartitionContainer(partitionId).getRecordStore(mapName, skipLoadingOnCreate);
+    }
+
+    @Override
     public RecordStore getExistingRecordStore(int partitionId, String mapName) {
         return getPartitionContainer(partitionId).getExistingRecordStore(mapName);
     }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/PartitionContainer.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/PartitionContainer.java
@@ -29,6 +29,7 @@ import com.hazelcast.spi.properties.HazelcastProperties;
 import com.hazelcast.util.ConcurrencyUtil;
 import com.hazelcast.util.ConstructorFunction;
 
+import java.util.Collection;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 
@@ -47,6 +48,15 @@ public class PartitionContainer {
             RecordStore recordStore = createRecordStore(name);
             recordStore.startLoading();
             return recordStore;
+        }
+    };
+
+    final ConstructorFunction<String, RecordStore> recordStoreConstructorSkipLoading
+            = new ConstructorFunction<String, RecordStore>() {
+
+        @Override
+        public RecordStore createNew(String name) {
+            return createRecordStore(name);
         }
     };
 
@@ -105,6 +115,10 @@ public class PartitionContainer {
         return maps;
     }
 
+    public Collection<RecordStore> getAllRecordStores() {
+        return maps.values();
+    }
+
     public int getPartitionId() {
         return partitionId;
     }
@@ -115,6 +129,11 @@ public class PartitionContainer {
 
     public RecordStore getRecordStore(String name) {
         return ConcurrencyUtil.getOrPutSynchronized(maps, name, this, recordStoreConstructor);
+    }
+
+    public RecordStore getRecordStore(String name, boolean skipLoadingOnCreate) {
+        return ConcurrencyUtil.getOrPutSynchronized(maps, name, this, skipLoadingOnCreate
+                ? recordStoreConstructorSkipLoading : recordStoreConstructor);
     }
 
     public RecordStore getRecordStoreForHotRestart(String name) {

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapReplicationOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapReplicationOperation.java
@@ -101,6 +101,12 @@ public class MapReplicationOperation extends AbstractOperation implements Mutati
         return new RecordReplicationInfo(key, mapServiceContext.toData(record.getValue()), info);
     }
 
+    private RecordStore getRecordStore(String mapName) {
+        final boolean skipLoadingOnRecordStoreCreate = true;
+        MapService mapService = getService();
+        MapServiceContext mapServiceContext = mapService.getMapServiceContext();
+        return mapServiceContext.getRecordStore(getPartitionId(), mapName, skipLoadingOnRecordStoreCreate);
+    }
 
     /**
      * Holder for raw IMap key-value pairs and their metadata.
@@ -137,13 +143,10 @@ public class MapReplicationOperation extends AbstractOperation implements Mutati
 
         private void applyState() {
             if (data != null) {
-                MapService mapService = getService();
-                MapServiceContext mapServiceContext = mapService.getMapServiceContext();
-
                 for (Entry<String, Set<RecordReplicationInfo>> dataEntry : data.entrySet()) {
                     Set<RecordReplicationInfo> recordReplicationInfos = dataEntry.getValue();
                     final String mapName = dataEntry.getKey();
-                    RecordStore recordStore = mapServiceContext.getRecordStore(getPartitionId(), mapName);
+                    RecordStore recordStore = getRecordStore(mapName);
                     recordStore.reset();
 
                     for (RecordReplicationInfo recordReplicationInfo : recordReplicationInfos) {
@@ -232,12 +235,9 @@ public class MapReplicationOperation extends AbstractOperation implements Mutati
 
 
         private void applyState() {
-            MapService mapService = getService();
-            MapServiceContext mapServiceContext = mapService.getMapServiceContext();
-
             for (Entry<String, List<DelayedEntry>> entry : delayedEntries.entrySet()) {
                 String mapName = entry.getKey();
-                RecordStore recordStore = mapServiceContext.getRecordStore(getPartitionId(), mapName);
+                RecordStore recordStore = getRecordStore(mapName);
                 WriteBehindStore mapDataStore = (WriteBehindStore) recordStore.getMapDataStore();
 
                 mapDataStore.reset();

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapReplicationOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapReplicationOperation.java
@@ -115,11 +115,16 @@ public class MapReplicationOperation extends AbstractOperation implements Mutati
     protected class MapReplicationStateHolder implements DataSerializable {
 
         protected Map<String, Set<RecordReplicationInfo>> data;
+        // propagates the information if the given record store has been already loaded with map-loaded
+        // if so, the loading won't be triggered again after a migration to avoid duplicate loading.
+        protected Map<String, Boolean> loaded;
 
         private void prepare(PartitionContainer container, int replicaIndex) {
             data = new HashMap<String, Set<RecordReplicationInfo>>(container.getMaps().size());
+            loaded = new HashMap<String, Boolean>(container.getMaps().size());
             for (Entry<String, RecordStore> entry : container.getMaps().entrySet()) {
                 RecordStore recordStore = entry.getValue();
+
                 MapContainer mapContainer = recordStore.getMapContainer();
                 MapConfig mapConfig = mapContainer.getMapConfig();
                 if (mapConfig.getTotalBackupCount() < replicaIndex) {
@@ -127,6 +132,7 @@ public class MapReplicationOperation extends AbstractOperation implements Mutati
                 }
                 MapServiceContext mapServiceContext = mapContainer.getMapServiceContext();
                 String mapName = entry.getKey();
+                loaded.put(mapName, recordStore.isLoaded());
                 // now prepare data to migrate records
                 Set<RecordReplicationInfo> recordSet = new HashSet<RecordReplicationInfo>(recordStore.size());
                 final Iterator<Record> iterator = recordStore.iterator();
@@ -148,6 +154,7 @@ public class MapReplicationOperation extends AbstractOperation implements Mutati
                     final String mapName = dataEntry.getKey();
                     RecordStore recordStore = getRecordStore(mapName);
                     recordStore.reset();
+                    recordStore.setPreMigrationLoadedStatus(loaded.get(mapName));
 
                     for (RecordReplicationInfo recordReplicationInfo : recordReplicationInfos) {
                         Data key = recordReplicationInfo.getKey();
@@ -163,13 +170,19 @@ public class MapReplicationOperation extends AbstractOperation implements Mutati
         @Override
         public void writeData(ObjectDataOutput out) throws IOException {
             out.writeInt(data.size());
-            for (Entry<String, Set<RecordReplicationInfo>> mapEntry : data.entrySet()) {
-                out.writeUTF(mapEntry.getKey());
-                Set<RecordReplicationInfo> recordReplicationInfos = mapEntry.getValue();
+            for (Entry<String, Set<RecordReplicationInfo>> dataEntry : data.entrySet()) {
+                out.writeUTF(dataEntry.getKey());
+                Set<RecordReplicationInfo> recordReplicationInfos = dataEntry.getValue();
                 out.writeInt(recordReplicationInfos.size());
                 for (RecordReplicationInfo recordReplicationInfo : recordReplicationInfos) {
                     out.writeObject(recordReplicationInfo);
                 }
+            }
+
+            out.writeInt(loaded.size());
+            for (Entry<String, Boolean> loadedEntry : loaded.entrySet()) {
+                out.writeUTF(loadedEntry.getKey());
+                out.writeBoolean(loadedEntry.getValue());
             }
         }
 
@@ -186,6 +199,12 @@ public class MapReplicationOperation extends AbstractOperation implements Mutati
                     recordReplicationInfos.add(recordReplicationInfo);
                 }
                 data.put(name, recordReplicationInfos);
+            }
+
+            int loadedSize = in.readInt();
+            loaded = new HashMap<String, Boolean>(loadedSize);
+            for (int i = 0; i < loadedSize; i++) {
+                loaded.put(in.readUTF(), in.readBoolean());
             }
         }
     }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
@@ -73,6 +73,9 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
     protected final MapKeyLoader keyLoader;
     // loadingFutures are modified by partition threads and could be accessed by query threads
     protected final Collection<Future> loadingFutures = new ConcurrentLinkedQueue<Future>();
+    // record store may be created with or without triggering the load
+    // this flag guards that the loading on create is invoked not more than once.
+    private boolean loadedOnCreate;
 
     public DefaultRecordStore(MapContainer mapContainer, int partitionId,
                               MapKeyLoader keyLoader, ILogger logger) {
@@ -81,10 +84,12 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
         this.logger = logger;
         this.keyLoader = keyLoader;
         this.recordStoreLoader = createRecordStoreLoader(mapStoreContext);
+        this.loadedOnCreate = false;
     }
 
     public void startLoading() {
-        if (mapStoreContext.isMapLoader()) {
+        if (mapStoreContext.isMapLoader() && !loadedOnCreate) {
+            loadedOnCreate = true;
             loadingFutures.add(keyLoader.startInitialLoad(mapStoreContext, partitionId));
         }
     }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
@@ -74,8 +74,11 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
     // loadingFutures are modified by partition threads and could be accessed by query threads
     protected final Collection<Future> loadingFutures = new ConcurrentLinkedQueue<Future>();
     // record store may be created with or without triggering the load
-    // this flag guards that the loading on create is invoked not more than once.
+    // this flag guards that the loading on create is invoked not more than once should the record store be migrated.
     private boolean loadedOnCreate;
+    // records if the record store has been loaded just before the migrations starts
+    // if so, the loading should NOT be started after the migration commit
+    private boolean loadedOnPreMigration;
 
     public DefaultRecordStore(MapContainer mapContainer, int partitionId,
                               MapKeyLoader keyLoader, ILogger logger) {
@@ -88,10 +91,28 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
     }
 
     public void startLoading() {
-        if (mapStoreContext.isMapLoader() && !loadedOnCreate) {
-            loadedOnCreate = true;
-            loadingFutures.add(keyLoader.startInitialLoad(mapStoreContext, partitionId));
+        if (logger.isFinestEnabled()) {
+            logger.finest("StartLoading invoked " + getStateMessage());
         }
+        if (mapStoreContext.isMapLoader() && !loadedOnCreate) {
+            if (!loadedOnPreMigration) {
+                if (logger.isFinestEnabled()) {
+                    logger.finest("Triggering load " + getStateMessage());
+                }
+                loadedOnCreate = true;
+                loadingFutures.add(keyLoader.startInitialLoad(mapStoreContext, partitionId));
+            } else {
+                if (logger.isFinestEnabled()) {
+                    logger.finest("Promoting to loaded on migration " + getStateMessage());
+                }
+                keyLoader.promoteToLoadedOnMigration();
+            }
+        }
+    }
+
+    @Override
+    public void setPreMigrationLoadedStatus(boolean loaded) {
+        loadedOnPreMigration = loaded;
     }
 
     @Override
@@ -101,6 +122,10 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
 
     @Override
     public void loadAll(boolean replaceExistingValues) {
+        if (logger.isFinestEnabled()) {
+            logger.finest("loadAll invoked " + getStateMessage());
+        }
+
         logger.info("Starting to load all keys for map " + name + " on partitionId=" + partitionId);
         Future<?> loadingKeysFuture = keyLoader.startLoading(mapStoreContext, replaceExistingValues);
         loadingFutures.add(loadingKeysFuture);
@@ -960,5 +985,11 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
             Record record = getRecordOrNull(toData(delayedEntry.getKey()), now, false);
             onStore(record);
         }
+    }
+
+    private String getStateMessage() {
+        return "on partitionId=" + partitionId + " on " + mapServiceContext.getNodeEngine().getThisAddress()
+                + " loadedOnCreate=" + loadedOnCreate + " loadedOnPreMigration=" + loadedOnPreMigration
+                + " isLoaded=" + isLoaded();
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/RecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/RecordStore.java
@@ -349,6 +349,17 @@ public interface RecordStore<R extends Record> {
     void startLoading();
 
     /**
+     * Informs this recordStore about the loading status of the recordStore that this store is migrated from.
+     * If the 'predecessor' has been loaded this record store should trigger the load again.
+     * Will be taken into account only if invoked before the startLoading method. Otherwise has no effect.
+     *
+     * This method should be deleted when the map's lifecycle has been cleaned-up. Currently it's impossible to
+     * pass additional state when the record store is created, thus this this state has to be passed in post-creation
+     * setters which is cumbersome and error-prone.
+     */
+    void setPreMigrationLoadedStatus(boolean loaded);
+
+    /**
      * Initialize the recordStore after creation
      */
     void init();

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapLoaderFailoverTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapLoaderFailoverTest.java
@@ -94,7 +94,7 @@ public class MapLoaderFailoverTest extends HazelcastTestSupport {
     }
 
     @Test(timeout = MINUTE)
-    // FIXED @Ignore //https://github.com/hazelcast/hazelcast/issues/6056
+    // FIXES https://github.com/hazelcast/hazelcast/issues/6056
     public void testLoadsAll_whenInitialLoaderNodeRemovedAfterLoading() throws Exception {
         Config cfg = newConfig("default", LAZY);
         HazelcastInstance[] nodes = nodeFactory.newInstances(cfg, 3);
@@ -108,6 +108,7 @@ public class MapLoaderFailoverTest extends HazelcastTestSupport {
 
         hz3.getLifecycleService().terminate();
         assertClusterSizeEventually(2, nodes[0]);
+
         map.loadAll(true);
 
         assertSizeEventually(MAP_STORE_ENTRY_COUNT, map);
@@ -142,7 +143,7 @@ public class MapLoaderFailoverTest extends HazelcastTestSupport {
     }
 
     @Test(timeout = MINUTE)
-    // FIXED https://github.com/hazelcast/hazelcast/issues/7959
+    // FIXES https://github.com/hazelcast/hazelcast/issues/7959
     public void testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups() throws Exception {
         PausingMapLoader<Integer, Integer> pausingLoader = new PausingMapLoader<Integer, Integer>(mapLoader, 5000);
 
@@ -181,7 +182,7 @@ public class MapLoaderFailoverTest extends HazelcastTestSupport {
         Config cfg = new Config();
         cfg.setGroupConfig(new GroupConfig(getClass().getSimpleName()));
         cfg.setProperty(GroupProperty.MAP_LOAD_CHUNK_SIZE.getName(), Integer.toString(BATCH_SIZE));
-        cfg.setProperty(GroupProperty.PARTITION_COUNT.getName(), "31");
+        cfg.setProperty(GroupProperty.PARTITION_COUNT.getName(), "13");
 
         MapStoreConfig mapStoreConfig = new MapStoreConfig()
                 .setImplementation(loader).setInitialLoadMode(loadMode);

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapLoaderFailoverTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapLoaderFailoverTest.java
@@ -29,7 +29,6 @@ import com.hazelcast.test.TestHazelcastInstanceFactory;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
 import org.junit.Before;
-import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
@@ -95,7 +94,7 @@ public class MapLoaderFailoverTest extends HazelcastTestSupport {
     }
 
     @Test(timeout = MINUTE)
-    @Ignore //https://github.com/hazelcast/hazelcast/issues/6056
+    // FIXED @Ignore //https://github.com/hazelcast/hazelcast/issues/6056
     public void testLoadsAll_whenInitialLoaderNodeRemovedAfterLoading() throws Exception {
         Config cfg = newConfig("default", LAZY);
         HazelcastInstance[] nodes = nodeFactory.newInstances(cfg, 3);
@@ -143,7 +142,7 @@ public class MapLoaderFailoverTest extends HazelcastTestSupport {
     }
 
     @Test(timeout = MINUTE)
-    @Ignore //https://github.com/hazelcast/hazelcast/issues/3796
+    // FIXED https://github.com/hazelcast/hazelcast/issues/7959
     public void testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups() throws Exception {
         PausingMapLoader<Integer, Integer> pausingLoader = new PausingMapLoader<Integer, Integer>(mapLoader, 5000);
 

